### PR TITLE
journald: harden against forward clock jumps before unclean shutdown

### PR DESCRIPTION
**Pull request by msizanoen1**
Try harder to inherit the sequence number and ID from the old journal file before rotating it away.

This helps the libsystemd journal file selection code make better decisions even in the face of massive incorrect forward clock jumps prior to an unclean shutdown.

### CODE REVIEW

File: src/journal/managed-journal-file.c [hunk 1]
```diff
@@ -538,6 +539,14 @@ int managed_journal_file_open_reliably(
         /* The file is corrupted. Rotate it away and try it again (but only once) */
         log_warning_errno(r, "File %s corrupted or uncleanly shut down, renaming and replacing.", fname);
 
+        if (!template) {
+                /* The file is corrupted and no template is specified. Try opening it read-only as the
+                 * template before rotating to inherit its sequence number and ID. */
+                managed_journal_file_open(-1, fname, O_RDONLY, file_flags, 0, compress_threshold_bytes, NULL,
+                                   mmap_cache, deferred_closes, NULL, &old_file);
```
Comment (yuwata) [inline src/journal/managed-journal-file.c]: Missing error handling, and spurious indentation.
(Please also fix the indentation below.)

**Comment by msizanoen1** [comment]
@yuwata Are there any remaining issues before this can be merged?

File: src/journal/managed-journal-file.c [hunk 2]
```diff
@@ -538,6 +539,17 @@ int managed_journal_file_open_reliably(
         /* The file is corrupted. Rotate it away and try it again (but only once) */
         log_warning_errno(r, "File %s corrupted or uncleanly shut down, renaming and replacing.", fname);
 
+        if (!template) {
+                /* The file is corrupted and no template is specified. Try opening it read-only as the
+                 * template before rotating to inherit its sequence number and ID. */
+                r = managed_journal_file_open(-1, fname, O_RDONLY, file_flags, 0, compress_threshold_bytes, NULL,
```
Comment (poettering) [inline src/journal/managed-journal-file.c]: i guess this doesn't matter much, but I think the open_flags should be specified as:

```c
…, (open_flags & ~(O_ACCMODE|O_CREAT|O_EXCL)) | O_RDONLY, …
```

**Comment by poettering** [comment]
lgtm

**Review by bluca — APPROVED** [review]

**Pull request merged**